PodSecurity message/check/fixture cleanups

liggitt opened the pull request:
#### What type of PR is this?

/kind cleanup
/kind feature

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/103373

Best reviewed commit-by-commit. Alternate commits are updating generated fixtures.

Address the following:
* Consistency of check IDs
* Make reason consistent (and consider exporting as a constant)
* Make details consistent
  * e.g. https://github.com/kubernetes/kubernetes/pull/103382#discussion_r662412946
* Ensure fixtures balance count/coverage (especially with variants covering initContainers)
  * Reduced from 3966 to 2553
* Comment blocks capturing the current policy (take it or leave it, but should be consistent)
* Usage of sets over slices (sets shuffle the order, and are less efficient when de-duplication isn't needed)

---

Review pass 1 (https://github.com/kubernetes/kubernetes/pull/103558#pullrequestreview-701536894):

checks:
- [x] allowPrivilegeEscalation
- [x] appArmorProfile
- [x] addCapabilities → capabilities_baseline
- [x] dropCapabilities → capabilities_restricted
- [x] hostNamespaces
- [x] hostPath → hostPathVolumes
- [x] hostPorts

---

Unreviewed:

checks:
- [x] hostProcess → windowsHostProcess
- [x] privileged
- [x] procMount
- [x] restrictedVolumes
- [x] runAsNonRoot
- [x] seccomp_restricted → seccompProfile_restricted
- [x] seccomp_baseline → seccompProfile_baseline
- [x] sysctls
- [x] selinux → seLinuxOptions

cleanup: drop field path from container visitor

limit warning evaluation/surfacing to requests that are not rejected by an enforce policy

reformat the warning message to read more naturally when admitting pods or pod-controllers

---

Follow-up:

* Avoid duplicate messages from overlapping checks (tracked in https://github.com/kubernetes/kubernetes/issues/103561)
  * restricted volumes complains about hostPath volumes already caught by baseline hostPath check
  * restricted capabilities complains about added non-default capabilities already caught by baseline capabilities check
  * restricted seccomp complains about unconfined seccomp profile already caught by baseline seccomp check


#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/assign @tallclair

liggitt committed:
PodSecurity: add message helper

liggitt committed:
PodSecurity: baseline capabilities: cleanup

Rename to capabilities_baseline
Add unit test exercising forbidden reason and details
Consolidate integration test fixtures

liggitt committed:
PodSecurity: baseline capabilities: regenerate files

liggitt committed:
PodSecurity: allowPrivilegeEscalation: cleanup

Make forbidden details more compact
Add unit test exercising forbidden message/details
Consolidate fixtures

liggitt committed:
PodSecurity: allowPrivilegeEscalation: regenerate files

liggitt committed:
PodSecurity: appArmorProfile: cleanup

Also allow  values
Add unit test exercising forbidden reason/detail
Clean up forbidden reason construction

liggitt committed:
PodSecurity: restricted capabilities: cleanup

Fix formatting of container names,
Add unit test for containers missing drop, containers with invalid adds
Consolidate integration test fixtures

liggitt committed:
PodSecurity: restricted capabilities: regenerate files

liggitt committed:
PodSecurity: hostNamespaces: cleanup

Use slice instead of set to accumulate errors
Add unit test to exercise message
Update docs to clarify undefined values are permitted

liggitt committed:
PodSecurity: hostPathVolumes: cleanup

Rename id to hostPathVolumes
Simplify message construction
Add unit test to exercise messages
Simplify integration test fixtures

liggitt committed:
PodSecurity: hostPathVolumes: regenerate files

liggitt committed:
PodSecurity: hostPorts: cleanup

Reformat message
Add unit test to exercise message/details

tallclair commented on staging/src/k8s.io/pod-security-admission/policy/check_capabilities_restricted.go:
```diff
@@ -64,36 +83,39 @@ func capabilitiesRestricted_1_22(podMetadata *metav1.ObjectMeta, podSpec *corev1
 	}
 	if len(containersAddingForbidden) > 0 {
-		forbiddenDetails = append(forbiddenDetails, fmt.Sprintf("containers %q must not add %q", containersAddingForbidden, forbiddenCapabilities.List()))
+		forbiddenDetails = append(forbiddenDetails, fmt.Sprintf(`containers %s must not include %s in securityContext.capabilities.add`, joinQuote(containersAddingForbidden), joinQuote(forbiddenCapabilities.List())))
```
nit: make this message consistent with the baseline capabilities details.

tallclair commented on staging/src/k8s.io/pod-security-admission/policy/check_apparmor.go:
```diff
@@ -56,31 +56,30 @@ func CheckAppArmorProfile() Check {
+			reason = "forbidden AppArmor profile"
+		} else {
+			reason = "forbidden AppArmor profiles"
+		}
```
nit: use pluralize

liggitt replied:
will do... didn't exist when I wrote this one  :)

tallclair reviewed (commented):
lgtm so far

liggitt committed:
PodSecurity: windowsHostProcess: cleanup

Rename to windowsHostProcess
Format reason/details
Add unit tests

liggitt committed:
PodSecurity: windowsHostProcess: regenerate files

liggitt committed:
PodSecurity: privileged: cleanup

liggitt committed:
PodSecurity: procMount: cleanup

liggitt committed:
PodSecurity: restrictedVolumes: cleanup

Updated forbidden reason/details
Added unit test to exercise all volume types
Consolidated fixtures

liggitt committed:
PodSecurity: restrictedVolumes: regenerate files

liggitt committed:
PodSecurity: runAsNonRoot: cleanup

Improve message and details
Add unit tests
Consolidate integration test fixtures

liggitt committed:
PodSecurity: runAsNonRoot: regenerate files

liggitt committed:
PodSecurity: seccompProfile_restricted: cleanup

Switch from field paths to container names in messages
Add unit tests for messages
Consolidate integration test fixtures

liggitt committed:
PodSecurity: seccompProfile_restricted: regenerate files

liggitt committed:
PodSecurity: seccompProfile_baseline: cleanup

Make messages consistent
Add unit tests for messages
Consolidate integration test fixtures
Rename to seccompProfile_baseline

liggitt committed:
PodSecurity: seccompProfile_baseline: regenerate files

liggitt committed:
PodSecurity: sysctls: cleanup

Add documentation and unit tests for messages

liggitt committed:
PodSecurity: seLinuxOptions: cleanup

rename to seLinuxOptions
make message consistent
add unit tests for message
consolidate integration test fixtures

liggitt committed:
PodSecurity: seLinuxOptions: regenerate files

liggitt committed:
PodSecurity: Drop field path from container visitor

liggitt committed:
limit warnings to requests that would otherwise succeed, reformat warning message

tallclair commented on staging/src/k8s.io/pod-security-admission/policy/check_restrictedVolumes.go:
```diff
@@ -89,16 +98,71 @@ func restrictedVolumes_1_0(podMetadata *metav1.ObjectMeta, podSpec *corev1.PodSp
+				badVolumeTypes.Insert("storageos")
+			default:
+				badVolumeTypes.Insert("unknown")
+			}
```
I suppose we shouldn't depend on PSP code? We might want to move https://github.com/kubernetes/kubernetes/blob/24a614cb4c5d8eaedffb4ad3b77e3c000d6e2b79/pkg/security/podsecuritypolicy/util/util.go#L81 to a common location.

liggitt replied:
I don't want to add any new dependencies on that code... looking forward to deleting it in 1.25

tallclair reviewed (approved):
/lgtm
